data/transactions/logic/assembler.go

```diff
@@ -1287,28 +1336,44 @@ func (ops *OpStream) assemble(fin io.Reader) error {
 				ok = false
 			}
 		}
+		if !ok {
+			// If the problem is only the version, it's useful to lookup the
+			// opcode from latest version, so we proceed with assembly well
+			// enough to report follow-on errors.  Of course, we still have to
+			// bail out on the assembly as a whole.
+			spec, ok = OpsByName[AssemblerMaxVersion][opstring]
+			if !ok {
+				spec, ok = keywords[opstring]
+			}
+			ops.errorf("%s opcode was introduced in TEAL v%d", opstring, spec.Version)
+		}
 		if ok {
 			ops.trace("%3d: %s\t", ops.sourceLine, opstring)
-			ops.RecordSourceLine()
-			if spec.Modes == runModeApplication {
+			ops.recordSourceLine()
+			if spec.Modes == modeApp {
 				ops.HasStatefulOps = true
 			}
-			args, returns := spec.Args, spec.Returns
-			if spec.Details.typeFunc != nil {
-				args, returns = spec.Details.typeFunc(ops, fields[1:])
+			args, returns := spec.Arg.Types, spec.Return.Types
+			if spec.OpDetails.refine != nil {
+				nargs, nreturns := spec.OpDetails.refine(ops.known, fields[1:])
+				if nargs != nil {
+					args = nargs
+				}
+				if nreturns != nil {
+					returns = nreturns
+				}
 			}
-			ops.checkStack(args, returns, fields)
+			ops.trackStack(args, returns, fields)
 			spec.asm(ops, &spec, fields[1:])
+			if spec.alwaysJumps() { // An unconditional branch deadens the following code
+				ops.known.deaden()
+			}
+			if spec.Name == "callsub" {
+				// since retsub comes back to the callsub, it is an entry point like a label
+				ops.known.label()
```

Comment:
@jannotti Half-baked question:  If you comment out the highlighted line, do you expect any unit tests to fail?

While trying to understand `TestTypeTracking`, I commented out the mentioned line thinking it'd cause at least one test failure in `gotestsum --  ./data/transactions/logic`, but it didn't.

I think I still don't fully grok `ProgramKnowledge` implications.  So, it's possible the behavior I'm describing is expected.  If easier to discuss live, happy to do so.

Reply:
I thought I had one that needed that behavior. I'll add one. The basic point is that if you do a callsub, you can not know anything about the stack on the following line because the subroutine might do arbitrary things and then return.

Reply:
Oh, cool.  Here's why nothing broke.  I had this test:
```
	// callsub also wipes our stack knowledge, this tests shows why: it's properly typed
	testProg(t, "callsub A; +; return; A: int 1; int 2; retsub", LogicVersion)
```
The goal was to show that `+` after callsub is allowed, because you don't know at that point what the subroutine is doing, so you must assume ints might be there.
With the line in question removed, everything after `callsub` is deadcode (because callsub is an unconditional jump).  Therefore, we're not even type checking after the `callsub`.  So I've added this test
```
	// but we do want to ensure we're not just treating the code after callsub as dead
	testProg(t, "callsub A; int 1; concat; return; A: int 1; int 2; retsub", LogicVersion,
		Expect{3, "concat arg 1 wanted..."})
```
It ensures that, although we don't know the condition of the stack after that `callsub`, we _do_ know that if we add an int to the stack, we can not call `concat` next.  This test will fail with the line you mentioned removed.

Reply:
@jannotti Thanks for the follow up here - Makes a lot of sense.  I had been thinking perhaps a test case is missing, but couldn't form the case in my mind.  Resolving.